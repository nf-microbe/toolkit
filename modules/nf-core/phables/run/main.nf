process PHABLES_RUN {
    tag "${meta.id}"
    label 'process_high'
    beforeScript "mkdir -p /tmp/phables_${meta.id}"
    containerOptions workflow.containerEngine == 'singularity' ?
        "--overlay /tmp/phables_${meta.id}":
        "--privileged"

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'docker://carsonjm/phables:1.4.0':
        'docker.io/carsonjm/phables:1.4.0' }"

    input:
    tuple val(meta) , path(fastq, stageAs: "reads/*")
    tuple val(meta2), path(graph)
    path phables_config
    path phables_db

    output:
    tuple val(meta2), path("${prefix}.phables.fasta.gz")        , emit: fasta
    tuple val(meta2), path("${prefix}.phables_complex_info.tsv"), emit: complex_info
    tuple val(meta2), path("${prefix}.phables_genome_info.tsv") , emit: genome_info
    tuple val(meta2), path("${prefix}.phables.log")             , emit: log
    path "versions.yml"                                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def is_compressed = graph.getExtension() == "gz" ? true : false
    def graph_name = is_compressed ? graph.getBaseName() : graph
    def phables_config_cp = phables_config ? "cp ${phables_config} ${meta.id}_config.yml" : ''
    def phables_config_cmd = phables_config ? "--configfile ${meta.id}_config.yml" : ''
    def conda_options = workflow.containerEngine == 'docker' ? "--no-use-conda" : ''
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    # decompress graph if compressed
    if [ "${is_compressed}" == "true" ]; then
        gzip -c -d ${graph} > ${graph_name}
    fi

    mkdir -p phables_out

    # create a copy of config
    ${phables_config_cp}

    phables \\
        run \\
        ${phables_config_cmd} \\
        --output phables_out \\
        --input ${graph_name} \\
        --reads reads \\
        --threads ${task.cpus} \\
        --prefix ${prefix} \\
        ${conda_options} \\
        ${args} || if [ -f phables_out/phables/resolved_component_info.txt ]; then true; else false; fi

    # check if phables outputs were created
    if [ ! -f phables_out/phables/resolved_paths.fasta ]; then
        touch phables_out/phables/resolved_paths.fasta
    fi

    # gzip resolved genomes
    gzip -c phables_out/phables/resolved_paths.fasta > ${prefix}.phables.fasta.gz
    mv phables_out/phables/resolved_component_info.txt ${prefix}.phables_complex_info.tsv
    mv phables_out/phables/resolved_genome_info.txt ${prefix}.phables_genome_info.tsv
    mv phables_out/phables.log ${prefix}.phables.log

    # remove extra files
    # rm -rf phables_out/

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        phables: \$(echo \$(phables -v 2>&1) | sed -n 's/phables, version //' ))
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def is_compressed = graph.getExtension() == "gz" ? true : false
    def graph_name = is_compressed ? graph.getBaseName() : graph
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    echo "" | gzip > ${prefix}.phables.fasta.gz
    touch ${prefix}.phables_complex_info.tsv
    touch ${prefix}.phables_genome_info.tsv
    touch ${prefix}.phables.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        phables: \$(echo \$(phables -v 2>&1) | sed -n 's/phables, version //' ))
    END_VERSIONS
    """
}
