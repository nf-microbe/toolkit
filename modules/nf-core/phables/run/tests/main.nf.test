nextflow_process {

    name "Test Process PHABLES_RUN"
    script "../main.nf"
    process "PHABLES_RUN"

    tag "modules"
    tag "modules_nfcore"
    tag "phables"
    tag "phables/install"
    tag "phables/run"

    setup {
        run("PHABLES_INSTALL") {
            script "../../install/main.nf"
        }
    }

    test("lambda - fastq - gfa") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [
                        file("${projectDir}/assets/test-datasets/genomics/lambda/illumina/fastq/SRR24085752_1.fastq.gz", checkIfExists: true),
                        file("${projectDir}/assets/test-datasets/genomics/lambda/illumina/fastq/SRR24085752_2.fastq.gz", checkIfExists: true)
                    ]
                ]
                input[1] = [
                    [ id:'test', single_end:false ], // meta map
                        file("${projectDir}/assets/test-datasets/genomics/lambda/illumina/gfa/SRR24085752.gfa.gz", checkIfExists: true)
                ]
                input[2] = [
                    file("${projectDir}/assets/configs/phables/phables_config.yml", checkIfExists: true),
                ]
                input[3] = PHABLES_INSTALL.out.phables_db
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.fasta,
                    process.out.info,
                    file(process.out.log[0][1]).name,
                    process.out.versions
                    ).match()
                }
            )
        }
    }

    test("lambda - fastq - gfa - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [
                        file("${projectDir}/assets/test-datasets/genomics/lambda/illumina/fastq/SRR24085752_1.fastq.gz", checkIfExists: true),
                        file("${projectDir}/assets/test-datasets/genomics/lambda/illumina/fastq/SRR24085752_2.fastq.gz", checkIfExists: true)
                    ]
                ]
                input[1] = [
                    [ id:'test', single_end:false ], // meta map
                        file("${projectDir}/assets/test-datasets/genomics/lambda/illumina/gfa/SRR24085752.gfa.gz", checkIfExists: true)
                ]
                input[2] = [
                    file("${projectDir}/assets/configs/phables/phables_config.yml", checkIfExists: true),
                ]
                input[3] = PHABLES_INSTALL.out.phables_db
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
