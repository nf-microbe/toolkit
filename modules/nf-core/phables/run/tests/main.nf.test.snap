{
    "lambda - fastq - gfa": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.phables.fasta.gz:md5,dba6802a454add1dc6f1dce9a5153d71"
                ]
            ],
            null,
            "test.phables.log",
            [
                "versions.yml:md5,36ef22678205006da92fe4354b0999b3"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-07T06:59:32.239063999"
    },
    "lambda - fastq - gfa - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.phables.fasta.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.phables_complex_info.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.phables_genome_info.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.phables.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,36ef22678205006da92fe4354b0999b3"
                ],
                "complex_info": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.phables_complex_info.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "fasta": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.phables.fasta.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "genome_info": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.phables_genome_info.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "log": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.phables.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,36ef22678205006da92fe4354b0999b3"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-07T15:35:00.184493848"
    }
}